File: pbmx_ffi/java/io/rmf/pbmx/payloads/RandomEntropyPayload.java
package io.rmf.pbmx.payloads;

import io.rmf.pbmx.BlockBuilder;
import io.rmf.pbmx.Mask;

public final class RandomEntropyPayload extends Payload {
    public RandomEntropyPayload(String name, Mask entropy) {
        this.name = name;
        this.entropy = entropy;
    }

    @Override
    public void addTo(BlockBuilder builder) {
        builder.randomEntropy(name, entropy);
    }

    private final String name;
    private final Mask entropy;
}

File: pbmx_droid/app/src/main/java/io/rmf/pbmx/payloads/RandomEntropyPayload.kt
package io.rmf.pbmx.payloads

import com.sun.jna.Pointer
import com.sun.jna.ptr.LongByReference
import io.rmf.pbmx.Mask
import io.rmf.pbmx.ffi.FFI
import io.rmf.pbmx.ffi.RawMask

class RandomEntropyPayload(handle: Pointer) : Payload(handle) {
    val name: String
    val entropy: Mask

    init {
        val length = LongByReference()
        var r = FFI.pbmx_unwrap_random_entropy(this.handle, null, length, null)
        assert(r == 0)

        val outName = ByteArray(length.value.toInt())
        val outEntropy = RawMask()
        r = FFI.pbmx_unwrap_random_entropy(this.handle, outName, length, outEntropy)
        assert(r != 0)
        this.name = outName.toString(Charsets.UTF_8)
        this.entropy = Mask(outEntropy)
    }
}

